### PR TITLE
If called from external with -Vi or lower, print command string

### PR DESCRIPTION
**Pull request by PaulWessel**
For simpler debugging from external calls, we echo out the module call string when **-Vi** is used.  Please give this a test, @seisman, to see if it works for PyGMT.

### CODE REVIEW

File: src/gmt_api.c [hunk 1]
```diff
@@ -11027,6 +11027,12 @@ int GMT_Call_Module (void *V_API, const char *module, int mode, void *args) {
 	else {	/* Call the function and pass back its return value */
 		gmt_manage_workflow (API, GMT_USE_WORKFLOW, NULL);		/* First detect and set modern mode if modern mode session dir is found */
 		gmtapi_check_for_modern_oneliner (API, module, mode, args);	/* If a modern mode one-liner we must switch run--mode here */
+		if (API->external && gmt_M_is_verbose (API->GMT, GMT_MSG_INFORMATION)) {
```
Comment (seisman) [inline src/gmt_api.c:11030]: I don't see any messages. Here you use `GMT_MSG_INFORMATION`, but later you use `GMT_MSG_DEBUG`. Does it make sense?

Comment (PaulWessel) [reply src/gmt_api.c:11030]: Makes no sense.  Cut/paste, now fixed.

Comment (PaulWessel) [reply src/gmt_api.c:11030]: If no commands are printed, perhaps the verbosity level is not set yet? API->external is set to true if GMT_SESSION_EXTERNAL is passed to GMT_Create_Session,a nd I think you do that.  So the only other way to fail (assuming you set -Vi somehow) is that it is not really set?  Parsing of the command has not happened yet, so need a way to do that.

Comment (PaulWessel) [reply src/gmt_api.c:11030]: Maybe try gmt set GMT_VERBOSE information before the command?

Comment (seisman) [reply src/gmt_api.c:11030]: Yes, `-Vi` doesn't work, but `gmt set GMT_VERBOSE` works.

However, it only outputs the arguments, not the module names.
```
pygmt-session [INFORMATION]: GMT_Call_Command string: 53c2e5b326d044899b91d46242b63ee3 -
pygmt-session [INFORMATION]: GMT_Call_Command string: -B -JM8i -R0/10/0/20 -Vi
pygmt-session [INFORMATION]: GMT_Call_Command string: 53c2e5b326d044899b91d46242b63ee3 -
pygmt-session [INFORMATION]: GMT_Call_Command string: @earth_relief_01d -Vi
pygmt-session [INFORMATION]: GMT_Call_Command string: 53c2e5b326d044899b91d46242b63ee3 -
pygmt-session [INFORMATION]: GMT_Call_Command string: -Ia/0.5,skyblue -Sskyblue -Vi
pygmt-session [INFORMATION]: GMT_Call_Command string: 53c2e5b326d044899b91d46242b63ee3 -
pygmt-session [INFORMATION]: GMT_Call_Command string: -N1/1,black -Vi
pygmt-session [INFORMATION]: GMT_Call_Command string: 53c2e5b326d044899b91d46242b63ee3 -
pygmt-session [INFORMATION]: GMT_Call_Command string: -A -Fales -Qg2 -Qt2 -Tg
pygmt-session [INFORMATION]: GMT_Call_Command string:
end [INFORMATION]: GMT_Call_Command string: 53c2e5b326d044899b91d46242b63ee3
```

Comment (weiji14) [reply src/gmt_api.c:11030]: Just tested this branch, and it seems to work, though I wouldn't say the 'GMT_Call_Command' string is very readable:

<details>

```bash
grdimage [INFORMATION]: GMT_Call_Command string: ~/.gmt/sessions/gmt_session.116530/=tiled_80_PO.000000 -R3.966666666666667/4.270833333333333/44/44.25 -I03s -rg -G@GMTAPI@-S-O-G-G-G-Y-000002 -fg -Co+n -N0
grdblend [INFORMATION]: Reading Data Table from file ~/.gmt/sessions/gmt_session.116530/=tiled_80_PO.000000
grdblend [INFORMATION]: Given domain implies x_inc = 0.000833333
grdblend [INFORMATION]: Given domain implies y_inc = 0.000833333
grdblend [INFORMATION]: Round-off patrol changed grid limit for ymin from 40.00000000000001 to 40.00000000000001
grdblend [INFORMATION]: Round-off patrol changed grid limit for ymax from 50.00000000000002 to 50.00000000000002
grdblend [INFORMATION]: Blend file ~/.gmt/server/earth/earth_relief/earth_relief_03s_g/N44E003.earth_relief_03s_g.nc in 3/4/44/45 with normal weight 1 [-900-300]
grdblend [INFORMATION]: Blend file ~/.gmt/server/earth/earth_relief/earth_relief_03s_g/N44E004.earth_relief_03s_g.nc in 4/5/44/45 with normal weight 1 [-900-300]
grdblend [WARNING]: File @N40E000.earth_relief_15s_p.nc has different increments (0.00416666666667/0.00416666666667) than the output grid (0.000833333333333/0.000833333333333) - must resample
grdblend [WARNING]: File @N40E000.earth_relief_15s_p.nc coordinates are phase-shifted w.r.t. the output grid - must resample
grdblend [INFORMATION]: Resample @N40E000.earth_relief_15s_p.nc via grdsample @N40E000.earth_relief_15s_p.nc -rg -I0.000833333333333/0.000833333333333 -R3.96666666667/4.27083333333/44/44.25 -G/tmp/grdblend_resampled_116530_2.nc -Vi -fg --GMT_HISTORY=readonly
grdblend [INFORMATION]: GMT_Call_Command string: @N40E000.earth_relief_15s_p.nc -rg -I0.000833333333333/0.000833333333333 -R3.96666666667/4.27083333333/44/44.25 -G/tmp/grdblend_resampled_116530_2.nc -Vi -fg --GMT_HISTORY=readonly
grdsample [INFORMATION]: Processing input grid
grdsample [INFORMATION]: Round-off patrol changed grid limit for ymin from 40.00000000000001 to 40.00000000000001
grdsample [INFORMATION]: Round-off patrol changed grid limit for ymax from 50.00000000000002 to 50.00000000000002
grdsample [INFORMATION]: Given domain implies x_inc = 0.000833333
grdsample [INFORMATION]: Given domain implies y_inc = 0.000833333
grdsample [INFORMATION]: Input  grid (0/10/40/50) n_columns = 2400 n_rows = 2400 dx = 0.00416666666667 dy = 0.00416666666667 registration = 1
grdsample [INFORMATION]: Output grid (3.96666666667/4.27083333333/44/44.25) n_columns = 366 n_rows = 301 dx = 0.000833333333315 dy = 0.000833333333333 registration = 0
grdsample [INFORMATION]: Reading grid from file ~/.gmt/server/earth/earth_relief/earth_relief_15s_p/N40E000.earth_relief_15s_p.nc
grdsample [INFORMATION]: All boundaries set via extended data.
grdsample [INFORMATION]: Writing grid to file /tmp/grdblend_resampled_116530_2.nc
grdblend [INFORMATION]: Round-off patrol changed grid limit for xmin from 3.966666666579721 to 3.966666666579721
grdblend [INFORMATION]: Round-off patrol changed grid limit for xmax from 4.270833333239721 to 4.270833333239721
grdblend [INFORMATION]: Blend file /tmp/grdblend_resampled_116530_2.nc in 0/10/40/50 with normal weight 1 [0-300]
grdblend [INFORMATION]: Processing input grids
grdblend [INFORMATION]: Round-off patrol changed grid limit for xmin from 3.966666666579721 to 3.966666666579721
grdblend [INFORMATION]: Round-off patrol changed grid limit for xmax from 4.270833333239721 to 4.270833333239721
grdblend [INFORMATION]: Processed row       0 of 301
grdblend [INFORMATION]: Processed row      10 of 301
grdblend [INFORMATION]: Processed row      20 of 301
grdblend [INFORMATION]: Processed row      30 of 301
grdblend [INFORMATION]: Processed row      40 of 301
grdblend [INFORMATION]: Processed row      50 of 301
grdblend [INFORMATION]: Processed row      60 of 301
grdblend [INFORMATION]: Processed row      70 of 301
grdblend [INFORMATION]: Processed row      80 of 301
grdblend [INFORMATION]: Processed row      90 of 301
grdblend [INFORMATION]: Processed row     100 of 301
grdblend [INFORMATION]: Processed row     110 of 301
grdblend [INFORMATION]: Processed row     120 of 301
grdblend [INFORMATION]: Processed row     130 of 301
grdblend [INFORMATION]: Processed row     140 of 301
grdblend [INFORMATION]: Processed row     150 of 301
grdblend [INFORMATION]: Processed row     160 of 301
grdblend [INFORMATION]: Processed row     170 of 301
grdblend [INFORMATION]: Processed row     180 of 301
grdblend [INFORMATION]: Processed row     190 of 301
grdblend [INFORMATION]: Processed row     200 of 301
grdblend [INFORMATION]: Processed row     210 of 301
grdblend [INFORMATION]: Processed row     220 of 301
grdblend [INFORMATION]: Processed row     230 of 301
grdblend [INFORMATION]: Processed row     240 of 301
grdblend [INFORMATION]: Processed row     250 of 301
grdblend [INFORMATION]: Processed row     260 of 301
grdblend [INFORMATION]: Processed row     270 of 301
grdblend [INFORMATION]: Processed row     280 of 301
grdblend [INFORMATION]: Processed row     290 of 301
grdblend [INFORMATION]: Processed row     300 of 301
grdblend [INFORMATION]: Processed row     301
grdblend [INFORMATION]: Referencing grid data to GMT_GRID memory location
grdblend [INFORMATION]: Set boundary condition for all edges: natural
grdblend [INFORMATION]: Set boundary condition for left   edge: natural
grdblend [INFORMATION]: Set boundary condition for right  edge: natural
grdblend [INFORMATION]: Set boundary condition for bottom edge: natural
grdblend [INFORMATION]: Set boundary condition for top    edge: natural
grdblend [INFORMATION]: Blended grid size of @GMTAPI@-S-O-G-G-G-Y-000002 is 366 x 301
grdblend [INFORMATION]: All nodes assigned values
grdimage [INFORMATION]: Set boundary condition for all edges: natural
grdimage [INFORMATION]: Set boundary condition for left   edge: natural
grdimage [INFORMATION]: Set boundary condition for right  edge: natural
grdimage [INFORMATION]: Set boundary condition for bottom edge: natural
grdimage [INFORMATION]: Set boundary condition for top    edge: natural
grdimage [INFORMATION]: Central meridian not given, default to 4.12
grdimage [INFORMATION]: Map scale is 1.6435 km per cm or 1:164350.
grdimage [INFORMATION]: Derive intensity grid from data grid
grdimage [INFORMATION]: Calling grdgradient with args -G@GMTAPI@-S-O-G-G-G-Y-000015 -A-45.0 -Nt1+a0 -R3.966666666666667/4.270833333333333/44/44.25 --GMT_HISTORY=readonly @GMTAPI@-S-I-G-G-G-N-000001
grdimage [INFORMATION]: GMT_Call_Command string: -G@GMTAPI@-S-O-G-G-G-Y-000015 -A-45.0 -Nt1+a0 -R3.966666666666667/4.270833333333333/44/44.25 --GMT_HISTORY=readonly @GMTAPI@-S-I-G-G-G-N-000001
grdgradient [INFORMATION]: Processing input grid
grdgradient [INFORMATION]: Set boundary condition for all edges: natural
grdgradient [INFORMATION]: Set boundary condition for left   edge: natural
grdgradient [INFORMATION]: Set boundary condition for right  edge: natural
grdgradient [INFORMATION]: Set boundary condition for bottom edge: natural
grdgradient [INFORMATION]: Set boundary condition for top    edge: natural
grdgradient [INFORMATION]:  Min Mean Max sigma intensities:
grdgradient [INFORMATION]: 	-0.720555859751	-0.00787141662076	0.890532427816	0.126627940389
grdgradient [INFORMATION]: Referencing grid data to GMT_GRID memory location
grdgradient [INFORMATION]: Set boundary condition for all edges: natural
grdgradient [INFORMATION]: Set boundary condition for left   edge: natural
grdgradient [INFORMATION]: Set boundary condition for right  edge: natural
grdgradient [INFORMATION]: Set boundary condition for bottom edge: natural
grdgradient [INFORMATION]: Set boundary condition for top    edge: natural
grdimage [INFORMATION]: Allocate and read data from file ~/.gmt/sessions/gmt_session.116530/=tiled_80_PO.000000
grdimage [INFORMATION]: Allocates memory and read intensity file
grdimage [INFORMATION]: project grid files
grdimage [INFORMATION]: Grid projection from size 364x301 to 364x301
grdimage [INFORMATION]: gmt_grd_project: Output grid extrema [73/1068] exceed extrema of input grid [73/715] due to resampling
grdimage [INFORMATION]: gmt_grd_project: Output grid clipped to input grid extrema
grdimage [INFORMATION]: Grid projection from size 366x301 to 366x301
grdimage [INFORMATION]: Reading CPT from File ~/gmt-install-dir/share/cpt/geo.cpt
grdimage [INFORMATION]: Auto-stretching CPT file geo to fit data range 73 to 715
grdimage [INFORMATION]: gmtsupport_validate_cpt: CPT hard hinge is outside actual data range - range adjusted to start at hinge 73 and below-hinge CPT ignored.
grdimage [INFORMATION]: Write CPT to File ~/.gmt/sessions/gmt_session.116530/gmt.2.cpt
grdimage [INFORMATION]: Save current CPT file to ~/.gmt/sessions/gmt_session.116530/gmt.2.cpt !
grdimage [INFORMATION]: Evaluate image pixel colors
grdimage [INFORMATION]: Basic z(x,y) with optional illumination and no PostScript colormasking.
grdimage [INFORMATION]: Creating 24-bit color image
PSL: Too many colors to make colormap - using 24-bit direct color instead.
PSL: DEFLATE compressed 328692 to 312437 bytes (4.9% savings at compression level 5)
```


</details>

**Comment by PaulWessel** [comment]
Yes, they kind of get buried among low-level statements.  Does it make sense for PyGMT to add some sort of debug-flag when you import, of is there no other place than passing a flag on the module call lines.

**Comment by joa-quim** [comment]
I don't need this from Julia, though it wouldn't harm, but it should not go into **-Vi**. Current attempt doesn't work either
```
julia> plot(rand(5,2), V=:i, par=(GMT_VERBOSE=:information,), Vd=1)
GMT [INFORMATION]: GMT_Call_Command string: -C @GMTAPI@-S-I-D-M-T-N-000000 ->@GMTAPI@-S-O-D-D-T-N-000001
gmtinfo [INFORMATION]: Processing input table data
gmtinfo [INFORMATION]: Reading Data Table from Input memory location via matrix
gmtinfo [INFORMATION]: Writing Data Table to memory reference supplied by pointer
        psxy  -JX12c/8c -Baf -BWSen --GMT_VERBOSE=information -R0.04/0.68/0.06/0.96 -Vi -P -K > C:\TMP\GMTjl_tmp.ps
psxy [INFORMATION]: Processing input table data
psxy [INFORMATION]: Auto-frame interval for x-axis (item 0): a0.2f0.1
psxy [INFORMATION]: Auto-frame interval for y-axis (item 0): a0.2f0.1
psxy [INFORMATION]: Map scale is 5.33333e-05 km per cm or 1:5.33333.
psxy [INFORMATION]: Duplicating data table from user matrix location
psxy [INFORMATION]: Plotting segment 0
```
Note that the inline argument ``--GMT_VERBOSE=information`` doesn't work. I had to previously call ``gmtset GMT_VERBOSE i``.

Probably the best place to implement this is in ``GMT_Create_Options()``, when
```
if (n_args_in == 0) {	/* Check if a single command line, if so break into tokens */
```
although here it doesn't know the module name, and cannot ever know it from MEX & Julia because we call ``GMT_Create_Options()`` directly.

**Comment by PaulWessel** [comment]
Do we hold onto this in master and remove the backport for now, @seisman?

**Comment by seisman** [comment]
Yes. "backport" removed.

**Comment by PaulWessel** [comment]
Some comments on the comments:

1. Yes, as mentioned one must need to do gmt set **GMT_HISTORY** _information_ since **--GMT_HISTORY**=_i_ is processed too late.
2. Plotting the comment without the module name is clearly not helpful, so I think that rules out _GMT_Create_Options_.
3. I forgot to add the module name before the options, now fixed.

Since this is to help debug, not for regular use by mere mortals, I do not think it is too much hardship to add a gmt set call to plant that setting before the main gmt commands.  See if this minor update is good enough for government work.

Note: This PR still does not address @weiji14 point that these lines are not so visible given all the other stuff that is printed, which is the same as @seisman point of adding a higher level special verbosity to just kick in these messages.

**Comment by seisman** [comment]
Tried this Python script:
```python
#!/usr/bin/env python
import pygmt

pygmt.config(GMT_VERBOSE='i')

fig = pygmt.Figure()
fig.basemap(region='g', projection='H15c', frame=['xaf', 'yaf', '+t"Global Map"'])
fig.coast(shorelines='1/0.5p')
fig.savefig("map.pdf")
```
And I get the full GMT commands by running the following command:
```bash 
$ python test.py 2>&1 | grep GMT_Call_Command | awk -F': ' '{print "gmt", $3}'
gmt figure 71485df0ed3b4d4983d7a2d6bc2a5445 -
gmt figure 71485df0ed3b4d4983d7a2d6bc2a5445 -
gmt basemap -Bxaf -Byaf -B+t"Global Map" -JH15c -Rg
gmt figure 71485df0ed3b4d4983d7a2d6bc2a5445 -
gmt coast -W1/0.5p
gmt figure 71485df0ed3b4d4983d7a2d6bc2a5445 -
gmt psconvert -A -Fmap -Qg2 -Qt2 -Tf
gmt end
gmt figure 71485df0ed3b4d4983d7a2d6bc2a5445
```
Seems not bad for developers.

@weiji14 What do you think?

**Comment by weiji14** [comment]
Looks much better! Probably not as ideal for users submitting bug reports (especially those on non-UNIX systems), but it should be enough for developers wanting to reproduce the commands. :+1: for me.

**Comment by PaulWessel** [comment]
Any concerns for Julia, @joa-quim ?

**Comment by seisman** [comment]
I'm wondering if we should move the messages to **-Vd** instead, as they're only useful for debugging purposes.

**Comment by weiji14** [comment]
> I'm wondering if we should move the messages to **-Vd** instead, as they're only useful for debugging purposes.

I suppose it should be `-Vd`. With the `grep` command, it should be easy enough to filter out the noisy debug messages (which was the original reason `-Vi` was used).

**Comment by joa-quim** [comment]
Well, I mentioned above some quirks from Julia, but nothing serious because I won't rely on this (which would partially fail for the reason referred).

The best, but worky, solution would be to have a function that takes the linked list of options and build an options string. The reverse of one that I don't remember the name.

**Comment by seisman** [comment]
Ping @PaulWessel to finalize this PR.

**Comment by PaulWessel** [comment]
> The best, but worky, solution would be to have a function that takes the linked list of options and build an options string. The reverse of one that I don't remember the name.

Do you mean _GMT_Create_Cmd_ ?  We have all of those in the API.

**Comment by PaulWessel** [comment]
To finalize, do you mean moving this to -Vd then, @seisman?

**Comment by seisman** [comment]
> To finalize, do you mean moving this to -Vd then, @seisman?

Yes to me, but I don't fully understand what @joa-quim said.

**Comment by joa-quim** [comment]
We have a function, ``GMT_Create_Options()``, that converts a string with the options to a linked list. What I meant is that the ideal would be to have a function to do the reverse. Then the passed options would be shown as a single string.

**Comment by PaulWessel** [comment]
And why is GMT_Create_Cmd not that function?

**Comment by joa-quim** [comment]
It looks it is, so the solution is to use it to rebuild the ``cmd``

**Comment by PaulWessel** [comment]
Which is what this PR does, no?

**Comment by joa-quim** [comment]
OK, I shut up. When I tested there were some issues but would need to repeat to learn again.

**Comment by PaulWessel** [comment]
OK @seisman I have moved it to -Vd and I hope this one is good to go.

File: src/gmt_api.c [hunk 2]
```diff
@@ -11034,6 +11034,12 @@ int GMT_Call_Module (void *V_API, const char *module, int mode, void *args) {
 	else {	/* Call the function and pass back its return value */
 		gmt_manage_workflow (API, GMT_USE_WORKFLOW, NULL);		/* First detect and set modern mode if modern mode session dir is found */
 		gmtapi_check_for_modern_oneliner (API, module, mode, args);	/* If a modern mode one-liner we must switch run--mode here */
+		if (API->external && gmt_M_is_verbose (API->GMT, GMT_MSG_DEBUG)) {
+			/* For externals only, print the equivalent command-line string under -Vi */
```
Comment (seisman) [inline src/gmt_api.c:11038]:
```suggestion
			/* For externals only, print the equivalent command-line string under -Vd */
```

**Review by seisman — APPROVED** [review]
It's good, except a typo.

**Pull request merged**